backend-rust/src/graphql_api/transaction_metrics.rs

```diff
@@ -0,0 +1,119 @@
+use std::sync::Arc;
+
+use async_graphql::{Context, Object, SimpleObject};
+use sqlx::postgres::types::PgInterval;
+
+use crate::graphql_api::{get_pool, ApiError, ApiResult, DateTime, MetricsPeriod, TimeSpan};
+
+#[derive(Default)]
+pub(crate) struct TransactionMetricsQuery;
+
+#[derive(SimpleObject)]
+struct TransactionMetrics {
+    /// Total number of transactions (all time).
+    last_cumulative_transaction_count: i64,
+    /// Total number of transactions in the requested period.
+    transaction_count: i64,
+    buckets: TransactionMetricsBuckets,
+}
+
+#[derive(SimpleObject)]
+struct TransactionMetricsBuckets {
+    /// The width (time interval) of each bucket.
+    bucket_width: TimeSpan,
+
+    /// Start of the bucket time period. Intended x-axis value.
+    #[graphql(name = "x_Time")]
+    x_time: Vec<DateTime>,
+
+    /// Total number of transactions (all time) at the end of the bucket period.
+    /// Intended y-axis value.
+    #[graphql(name = "y_LastCumulativeTransactionCount")]
+    y_last_cumulative_transaction_count: Vec<i64>,
+
+    /// Total number of transactions within the bucket time period. Intended
+    /// y-axis value.
+    #[graphql(name = "y_TransactionCount")]
+    y_transaction_count: Vec<i64>,
+}
+
+#[Object]
+impl TransactionMetricsQuery {
+    async fn transaction_metrics(
+        &self,
+        ctx: &Context<'_>,
+        period: MetricsPeriod,
+    ) -> ApiResult<TransactionMetrics> {
+        let pool = get_pool(ctx)?;
+
+        let last_cumulative_transaction_count = sqlx::query_scalar!(
+            "SELECT cumulative_num_txs FROM blocks ORDER BY height DESC LIMIT 1"
+        )
+        .fetch_one(pool)
+        .await?;
+
+        let interval: PgInterval = period
+            .as_duration()
+            .try_into()
+            .map_err(|e| ApiError::DurationOutOfRange(Arc::new(e)))?;
```

Comment:
Just curious: why do we need `arc` on the error?

Reply:
I was curious as well - the reason seems to be that `async_graphql` requires the return types of the graphql queries to be `Clone`, which means the error type must be clone. However some of the errors in the variants of the error enum are *not* `Clone`, so they are wrapped in `Arc` to make them `Clone`.